### PR TITLE
Deprecate `ReconnectMode` from container-loader, re-create in container-definitions

### LINKED ISSUE #9166: Move `ReconnectMode` to container-definitions
As part of the plan to use `ReconnectMode` enum as an argument in the reworked `setAutoReconnect()` (see https://github.com/microsoft/FluidFramework/issues/9167), `ReconnectMode` will be moved to container-definitions. The new `ReconnectMode` is planned to be defined as:
```ts
export enum ReconnectMode {
    Never = "Never",
    Disabled = "Disabled",
    Enabled = "Enabled",
    Custom = "Custom"
}
```
The `Custom` enum member will be introduced to enable the developer to use their own callback function to determine when the container should and should not be reconnected (see https://github.com/microsoft/FluidFramework/issues/9167 for more details).

Tasks:
- [ ] Deprecate `ReconnectMode` from `container-loader` and re-create in `container-definitions`
- [ ] Replace all usages of `ReconnectMode` with the new version
- [ ] Remove deprecated `ReconnectMode` after two releases

### PR DESCRIPTION
**Pull request by scottn12**
This PR starts the process of moving the `ReconnectMode` enum from container-loader to contianer-definitions. This deprecates the version in container-loader and re-creates it in `container-loader` (with an additional enum member).

Context: https://github.com/microsoft/FluidFramework/issues/9166

### CODE REVIEW

File: common/lib/container-definitions/src/loader.ts [hunk 1]
```diff
@@ -148,6 +148,16 @@ export namespace ConnectionState {
  */
 export type ConnectionState = ConnectionState.Disconnected | ConnectionState.Connecting | ConnectionState.Connected;
 
+/**
+ * Accepted reconnection modes to be used with `setAutoReconnect()`
+ */
+export enum ReconnectMode {
```
Comment (anthony-murphy) [inline common/lib/container-definitions/src/loader.ts:154]: why are we moving this to common? that seems really odd.

Comment (anthony-murphy) [reply common/lib/container-definitions/src/loader.ts:154]: i don't think these changes should be coupled

**Comment by vladsud** [comment]
It would be great to have some mini-design review here, as I'm not sure what you are trying to achieve and why.
I glanced at https://github.com/microsoft/FluidFramework/issues/9167, but I do not see an articulation of what problem you are trying to solve. Defined through "user can" kind of description.

The other comment I have - if we are touching these areas, then I'd love us to also do it right. IMHO that means moving it to loader and ensuring that policies apply to all containers created under same loader. This will ensure that any "guest" component (running in another container created by main container) have exactly same policy applied. There will be exceptions (like summarizer client), but I think APIs like that, forceReadonly, and couple others are all candidates to be moved to "context" object (which is loader, I do not think we want another context object here).

File: common/lib/container-definitions/src/loader.ts (hunk 1 above)
Comment (scottn12) [reply common/lib/container-definitions/src/loader.ts:154]: The motivation behind moving it is based on https://github.com/microsoft/FluidFramework/issues/9167. To summarize, we want to rework `setAutoReconnect()` and have it use `ReconnectMode` as the first argument in the function. We also want to expose `setAutoReconnect()` so the developer will need to have access to `ReconnectMode`.

Comment (anthony-murphy) [reply common/lib/container-definitions/src/loader.ts:154]: i wouldn't use an enum here. we should just use `export type ReconnectMode= ...` and let the consumer bring their own strings

Comment (anthony-murphy) [reply common/lib/container-definitions/src/loader.ts:154]: would be great it the spec were in the design item: https://github.com/microsoft/FluidFramework/issues/8745

Comment (anthony-murphy) [reply common/lib/container-definitions/src/loader.ts:154]: if you just add the type. you could keep the enum at the loader layer, probably don't need to export it

File: packages/loader/container-loader/src/contracts.ts [hunk 2]
```diff
@@ -21,6 +21,11 @@ import {
     ISignalMessage,
 } from "@fluidframework/protocol-definitions";
 
+/**
+ * @deprecated 0.58, Will be removed in 0.60
+ * Moved to {@link @fluidframework/container-definitions}
+ * See https://github.com/microsoft/FluidFramework/issues/9166 for context
+ */
 export enum ReconnectMode {
```
[the remark this answers is not in the document]
Comment (anthony-murphy) [reply packages/loader/container-loader/src/contracts.ts:29]: this is good. i would keep this, and only add a type to the definitions

File: common/lib/container-definitions/src/loader.ts [hunk 3]
```diff
@@ -148,6 +148,16 @@ export namespace ConnectionState {
...
+    Never = "Never",
+    Disabled = "Disabled",
```
Comment (anthony-murphy) [inline common/lib/container-definitions/src/loader.ts:156]: what's the difference between Never and disabled

**Comment by anthony-murphy** [comment]
> > It would be great to have some mini-design review here, as I'm not sure what you are trying to achieve and why.
> 
> @vladsud @anthony-murphy, I would be happy to have a design review to go over these topics. We had a smaller design review meeting separately as part of #8745, but I think it may be a good idea to have another with a larger audience now that we have more of a solidified plan.

i much prefer async review in an issue, and only to hold a meeting if we need it. we are also open source, so doing our reviews in issue and in the open is important

File: common/lib/container-definitions/src/loader.ts (hunk 1 above)
Comment (scottn12) [reply common/lib/container-definitions/src/loader.ts:154]: Are there any other advantages of using a type vs an enum in this situation (other than not having to move the current `ReconnectMode` enum)?

Comment (anthony-murphy) [reply common/lib/container-definitions/src/loader.ts:154]: enums emit javascript code. types do not. i'd like our definition packages to eventually be types only

**Comment by scottn12** [comment]
Closed due to updated solution design

**Pull request closed without merging**